Add `rrules` for Interpolations types and tests

SomTambe opened the pull request:
I have added the homography struct and `rrule`s. I am currently in the process of adding correct tests using `ChainRulesTestUtils.test_rrule`.

SomTambe commented:
What would be the correct tangents for the primal `Interpolations.BSplineInterpolation`?
From the [docs](https://github.com/JuliaMath/Interpolations.jl/blob/31f6d54fb8d9eab0511a1c50182025b0a448c9f1/src/b-splines/b-splines.jl#L47-L68), this looks like the right tangents for the primal:
```julia
Tangent{Interpolations.BSplineInterpolation}(;coefs = rand(3,3), parentaxes = typeof((1:3, 1:3)), it=BSpline(Linear()))
```
Is this right? @DhairyaLGandhi @johnnychen94 ?

johnnychen94 commented on src/geometry/adjoints.jl:
```diff
@@ -0,0 +1,52 @@
+using Base: Tuple
+function ChainRulesCore.rrule(::typeof(Interpolations.BSplineInterpolation), T, A, it, axs)
+    y = Interpolations.BSplineInterpolation(T, A, it, axs)
+    return (y, Δ -> begin
+        (NoTangent(), NoTangent(), Δ, NoTangent(), NoTangent())
+    end)
+end
```
For constructors and functors, I don't think can dispatch like this: it should be

```julia
function ChainRulesCore.rrule(::Interpolations.BSplineInterpolation, T, A, it, axs)
```

See also: https://juliadiff.org/ChainRulesCore.jl/dev/writing_good_rules.html#Structs:-constructors-and-functors

DhairyaLGandhi replied:
Stylistically, CR does suggest a `return` statement, but you don't need to wrap it in a tuple. Also you may want to write the `function $fn_to_be_differentiated_pullback`  syntax.

SomTambe replied:
> For constructors and functors, I don't think can dispatch like this: it should be

Yeah exactly, all the `rrule`s were written the same way before I started getting some errors during testing. I'll modify them again.

johnnychen94 commented on src/geometry/adjoints.jl:
```diff
@@ -0,0 +1,52 @@
+    end)
+end
+
+Zygote.@nograd Interpolations.tweight
+Zygote.@nograd Interpolations.tcoef
```
Also transform this into ChainRules' language because ChainRules doesn't recognize Zygote.

DhairyaLGandhi replied:
I think the `@non_diff` notation is slightly verbose and Zygote preferentially chooses its own rules which can be better optimised in cases, so it's alright, but you should add the CRC method in here as well.

johnnychen94 reviewed (commented):
This PR is not atomic and you can still separate it into multiple parts. For example: 1) add missing `rrule`s for Interpolations, and 2) add homography matrix type and verify AD works on it.

DhairyaLGandhi reviewed (commented):
Stylistically again: don't add empty files or references to it.

DhairyaLGandhi commented:
Why would the coefs be `rand`?

SomTambe commented:
> the `coefs` field holds the interpolation coefficients. Depending on prefiltering, these may or may not be the same as the supplied array of interpolant values.

For my case where I just extended it to `m = rand(3,3)`, I thought they would be the same as the array `m`.

SomTambe commented:
For some reason I feel it is a bad idea to write tests for rules using ChainRulesTestUtils. Here's what I've been trying to do. I have been testing for the constructor [`Interpolations.BSplineInterpolation`](https://github.com/JuliaMath/Interpolations.jl/blob/31f6d54fb8d9eab0511a1c50182025b0a448c9f1/src/b-splines/b-splines.jl#L69), should not have been difficult, but I now have no idea at all about how to resolve the Inexact error that I've been getting:
```julia
function ChainRulesCore.rrule(::Type{Interpolations.BSplineInterpolation{T, N, TA, IT, TAX}}, A, axs, it) where {T, N, TA, IT, TAX}
    y = Interpolations.BSplineInterpolation{T, N, TA, IT, TAX}(A, axs, it)
    function bspline_construct_pullback(Δy)
        return NoTangent(), Δy, NoTangent(), NoTangent()
    end
    return y, bspline_construct_pullback
end

julia> ty = Interpolations.BSplineInterpolation{Float64, 2, typeof(m), BSpline{Linear{Throw{OnGrid}}}, typeof(axs)} # axs = (Base.OneTo(3), Base.OneTo(3)), can also use (1:3, 1:3).
julia> test_rrule(ty, m, axs, BSpline(Linear())); # m = rand(3, 3)
test_rrule: Interpolations.BSplineInterpolation{Float64, 2, Matrix{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}} on Matrix{Float64},Base.OneTo{Int64},Base.OneTo{Int64},BSpline{Linear{Throw{OnGrid}}}: Error During Test at /home/somvt/.julia/packages/ChainRulesTestUtils/AX7fv/src/testers.jl:227
  Got exception outside of a @test
  InexactError: Int64(2.99)
...
Test Summary:                                                                                                                                                                                                                                   | Pass  Error  Total
test_rrule: Interpolations.BSplineInterpolation{Float64, 2, Matrix{Float64}, BSpline{Linear{Throw{OnGrid}}}, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}} on Matrix{Float64},Base.OneTo{Int64},Base.OneTo{Int64},BSpline{Linear{Throw{OnGrid}}} |    3      1      4
ERROR: Some tests did not pass: 3 passed, 0 failed, 1 errored, 0 broken.
```

johnnychen94 commented:
Some tough words to spell but it's really nothing personal.

It is a bad idea to not write tests because you're too lazy to figure out the deeply buried issue. People always find excuses to not write tests, people always find them bitten later because of not writing tests. I don't need to prove this again to you and really I don't want to say it for every PR you've submitted.

JSoC students are potential package developers and not package users. The JSoC event is not asking you to report issues, is asking you to fix issues.

So please, be patient to dig into the details and do a large portion of the work in the background before raising a new issue. Can I at least see you independently finish one small PR without any feedback?

I'm saying this very seriously because you still haven't show your ability to work independently even if we've gone through three merged PRs. So please take this as a warning.

SomTambe commented:
Added rrules, and tests for the following rrules using `FiniteDifferences`:

- [x] Interpolations.BSplineInterpolation (2)
- [x] Interpolations.Extrapolation
- [x] Interpolations.copy_with_padding
- [x] Interpolations.FilledExtrapolation

Removed homography structs. I Will put them as a separate PR along with adjoints.

SomTambe commented:
The `Params not defined` error seems like a Zygote issue, but cannot reproduce it locally. Let me check if changing to the latest version of Zygote works.

SomTambe commented:
The `Interpolations.FilledExtrapolation` constructor gradient is failing because FiniteDifferences gets messed up if given an Interpolation object as the input of a function to calculate the gradient of. It can be resolved if the fdm tests are modified to be something like this:
```julia
fd = grad(central_fdm(5,1), x->begin
                                   Interpolations.FilledExtrapolation(ty(x, axs, BSpline(Linear())), 0.0) |> sum
                               end, 
                               m)
```
Although these would be the correct tests numerically and these outputs are expected even with the existing tests, I believe this is not the correct way to test since the Interpolations object is not being tested for the gradient at all.

DhairyaLGandhi commented on src/geometry/adjoints.jl:
```diff
@@ -0,0 +1,58 @@
+Zygote.@nograd Interpolations.tweight
+Zygote.@nograd Interpolations.tcoef
+
+function ChainRulesCore.rrule(::typeof(Interpolations.copy_with_padding), T, A, it)
```
Does this need an rrule or can it be a nograd?

SomTambe replied:
I think we can do a nograd, but it will collide with the fact that we plan to use the image gradients ∂L/∂X at a later part of our project.

DhairyaLGandhi replied:
Yeah, just check what the impact is and we can take a call then

DhairyaLGandhi commented on src/geometry/adjoints.jl:
```diff
@@ -0,0 +1,58 @@
+function ChainRulesCore.rrule(::Interpolations.BSplineInterpolation, T, A, it, axs)
```
Do you need an rrule for this method if you already have one for the type constructor? What happens if you remove it?

SomTambe replied:
It actually does not make a difference! I can do away with it to be honest, it always falls back to the constructor adjoints. Should I do that? I'll add it again if we collude with it at some latter part.

DhairyaLGandhi commented on test/geometry/adjoints.jl:
```diff
@@ -0,0 +1,56 @@
+
+@testset "Interpolations.BSplineInterpolation constructor gradient" begin
+    zg = Zygote.gradient((x,y,z)->ty(x,y,z) |> sum, m, axs, BSpline(Linear()))
+    fd = grad(central_fdm(5,1), x->begin
```
Remove the begin and move the anonymous function to the next line instead?

johnnychen94 commented:
Please try your best to push a version that CI can actually pass. If it doesn't work somewhere, place the `@test` with `@test_broken`.

johnnychen94 commented on src/geometry/adjoints.jl:
```diff
@@ -0,0 +1,58 @@
+    return y, filledextra_const_pb
+end
+
+function ChainRulesCore.rrule(::Type{SVector{N, T}}, x...) where {N, T}
+    y = SVector{N,T}(x...)
+    function svector_const_pb(Δy)
+        return NoTangent(), Δy
+    end
+    return y, svector_const_pb
+end
+
+function ChainRulesCore.rrule(::CartesianIndices, t::Tuple)
+    y = CartesianIndices(t)
+    function cartesianindices_pb(Δy)
+        return NoTangent(), Δy
+    end
+    return y, cartesianindices_pb
+end
```
Also, add some test for these two rules, check the types. Without these, I believe Zygote will give `NamedTuple`.

johnnychen94 replied:
Also, add some test for these two rules, check the types. Without these, I believe Zygote will give `NamedTuple`.

johnnychen94 replied:
Also, add some test for these two rules, check the types. Without these, I believe Zygote will give `NamedTuple`.

johnnychen94 replied:
Also, add some test for these two rules, check the types. Without these, I believe Zygote will give `NamedTuple`.

johnnychen94 reviewed (commented):
Nothing suspicious to me. I won't stand in the way of merging this. But a PR is expected to be minimal:

- [x] resolve the comments
- [x] remove `Images`, `CoordinateTransformations`, `OffsetArrays`, `ColorVectorSpace` from the dependency, use `ImageCore@0.9` instead.

Style suggestion:

- [x] don't use `|>` in the library codebase. This is only for personal convenience purposes.

SomTambe commented:
> remove Images, CoordinateTransformations, OffsetArrays, ColorVectorSpace from the dependency, use ImageCore@0.9 instead.

I need `ColorVectorSpace` since I use the `\cdot` operation in one of my rrules. I also need CoordinateTransformations since I am defining the Homography struct to be a `<: CoordinateTransformations.Transformation`. So I think doing away with only the Images and OffsetArrays libraries should be good, I'll keep the rest two.

johnnychen94 commented on .github/workflows/ci.yml:
```diff
@@ -3,6 +3,7 @@ on:
   push:
     branches:
       - main
+      - *
```
This is invalid syntax.

```suggestion
    branches:
      - main
```

CI not running in #16 is because there are conflicts. You have to resolve conflicts first.

SomTambe replied:
yeah sure. once this PR gets merged, I'll change the base to main and resolve the conflicts, and then run it again.

SomTambe commented:
All tests passed, ig we're gtg!

DhairyaLGandhi commented on src/DiffImages.jl:
```diff
@@ -1,16 +1,21 @@
+include("geometry/adjoints.jl")
 
-end # module
+end
```
Check your editor settings please, this should be in the Linux file endings

SomTambe replied:
Oh, so should I revert the changes? Should I keep `# module`?

DhairyaLGandhi replied:
Do you see the little red circle with a cross in the diff? That's what I was referring to. No harm keeping the comment, no hard removing. File endings should be neat though.

SomTambe replied:
I had to go to the diff to check out the red cross, did not know about this. My editor defaults to clrf I think so. Should I insert newlines at the end of files?

DhairyaLGandhi replied:
Your editor must have some settings to control this, change that, save the files and commit

SomTambe replied:
Done.

DhairyaLGandhi reviewed (approved):
Took a while, but kudos for sticking with it! Make sure to squash and merge.

DhairyaLGandhi commented on Project.toml:
```diff
@@ -5,8 +5,35 @@ version = "0.1.0"
+[extras]
+ImageCore = "a09fc81d-aa75-5fe9-8630-4744c3626534"
+FiniteDifferences = "26cc04aa-876d-5657-8c51-4c34ba976000"
+Flux = "587475ba-b771-5e3f-ad9e-33799f191a9c"
```
You don't need the packages that are direct deps in extras

SomTambe replied:
But I need to put the test dependencies that I have mentioned in the [targets] section to be in extras right, without that, it errors.

DhairyaLGandhi replied:
Are you sure?

SomTambe replied:
Yeah I just tried it out again, it does not work. Whatever you want in your test deps needs to be in extras as well.

DhairyaLGandhi replied:
What happens if you remove them from both?

johnnychen94 replied:
There might be a misunderstanding here. Every entry that you listed in [deps] section is automatically also a test dependency, which you don't need to explicitly declare again as [targets.test], and thus you don't need them added to [extra] section.

SomTambe replied:
Ahhh this is right, I've updated it. Now should do the work :p

SomTambe commented:
Made the changes that you guys requested. We're good to go now, and I'm merging this PR.
